packages/components/src/components/Calculator/index.tsx

```diff
@@ -1,158 +1,237 @@
-import {
-  CalculatorAction,
-  CalculatorState,
-  calculatorReducer,
-  hasSameCalculator,
-  initialCalculatorState,
-} from "./calculatorReducer.js";
-import { CalculatorUI } from "./calculatorUI.js";
+  const hashString = useMemo(() => calculator.hashString, [calculator]);
+  const [prevHashString, setPrevHashString] = useState<string | undefined>(
+    cachedState?.hashString
+  );
+
+  // This effect does two things:
+  // - when the calculator is updated, it resets the form
+  // - also, on initial render and on calculator updates, it calculates all input values (which will trigger calculator result calculation)
+  // In both cases, it updates `prevHashString`, which tracks the calculator identity.
+  useEffect(() => {
+    if (prevHashString === hashString) {
+      return;
+    }
+
+    if (prevHashString) {
+      // Calculator identity has changed. Let's reset the form because there might be new inputs or new defaults.
+      // (In the future, we could do something more complicated here, for example check if any form fields were touched.)
+      form.reset(getFormValues(calculator));
+    }
+
+    // Calculator has updated (or this component just mounted). Let's take all input codes and run them.
+    const processAllFieldCodes = async () => {
+      const formValues = form.getValues();
+      const newInputValues: typeof inputValues = {};
+      for (const input of calculator.inputs) {
+        const name = input.name;
+        const fieldValue = formValues[name];
+        const inputValueResult = await runSquiggleCode(
+          fieldValueToCode(name, calculator, fieldValue),
+          environment
+        );
+        newInputValues[name] = inputValueResult;
+      }
+      // All values are updated simultaneously, to avoid too many rerenders.
+      setInputValues(newInputValues);
+    };
+    processAllFieldCodes();
+
+    setPrevHashString(hashString);
+  }, [calculator, environment, hashString, prevHashString, form]);
+
+  // Update input value if input code has changed.
+  useEffect(() => {
+    const subscription = form.watch(async (formValues, { name }) => {
+      if (name === undefined) return;
+      const fieldValue = formValues[name];
+      if (fieldValue === undefined) return;
+
+      const valueResult = await runSquiggleCode(
+        fieldValueToCode(name, calculator, fieldValue),
+        environment
+      );
+      setInputValues((inputValues) => ({
+        ...inputValues,
+        [name]: valueResult,
+      }));
+    });
+
+    return () => subscription.unsubscribe();
+  }, [calculator, environment, form]);
+
+  // Back up calculator state outside of this component, since calculator can disappear on code changes
+  // (for example, on short syntax errors when autorun is enabled), and we don't want to lose user input
+  // or calculator output when calculator component is rendered again.
+  const backupState = useCallback(() => {
```

Comment:
``backupState`` means the same thing to me as ``updateCachedState``. I could see just calling this, ``_updateCachedState`` or something else that makes the equivalency clearer.

Reply:
Also, it might well make sense to rename "updateCachedState" to "updateCacheState" (remove the d)

Reply:
> Also, it might well make sense to rename "updateCachedState" to "updateCacheState" (remove the d)

Renamed to `updateSavedState`, now it matches the hook name

> `backupState` means the same thing to me as `updateCachedState`. I could see just calling this, `_updateCachedState` or something else that makes the equivalency clearer.

Done, thanks, I didn't like the name either.